subsystems/networking/bluetooth_linux.go

```diff
@@ -89,28 +107,96 @@ func (n *Networking) initializeBluetoothService(deviceName string, characteristi
-// When that happens, we get instantly disconnected due to OS security on the client (where full pairing is required.)
-func (n *Networking) writeBTDisableDiscovery(ctx context.Context) error {
-	contents := BTDiscoveryContentsDisable
-	if n.Config().DisableBTProvisioning {
-		contents = BTDiscoveryContentsDefault
+type basicAgent struct {
+	conn   *dbus.Conn
+	logger logging.Logger
+}
+
+// TODO(APP-8073) This is a workaround, rejecting "just-works" pairing requests generated by pop-up requests during provisioning
+// so that ONLY pairing from the device's OS works, (since the NoInputNoOutput mode we register normally doesn't even query our agent.)
+// Only pairing from there appears to properly set up bluetooth tethering, and people can't easily re-pair.
+func (b *basicAgent) RequestAuthorization(device dbus.ObjectPath) *dbus.Error {
+	b.logger.Infof("rejecting just-works bluetooth pair attempt for %s, please initiate pairing from your mobile device settings", device)
+	return &dbus.Error{Name: "org.bluez.Error.Rejected"}
+}
+
+func (n *Networking) enablePairing(deviceName string) error {
+	conn, adapter, err := getBluetoothDBus()
+	if err != nil {
+		return err
 	}
 
-	isNew, err := utils.WriteFileIfNew(BTDiscoveryFilepath, []byte(contents))
+	err = adapter.SetProperty("org.bluez.Adapter1.Alias", dbus.MakeVariant(deviceName))
 	if err != nil {
-		return fmt.Errorf("writing %s: %w", BTDiscoveryFilepath, err)
+		return errw.Wrap(err, "setting bluetooth alias")
 	}
 
-	if isNew {
-		n.logger.Infof("Updated %s to: %q", BTDiscoveryFilepath, contents)
-		// restart bluetooth to apply changes
-		cmd := exec.CommandContext(ctx, "systemctl", "restart", "bluetooth")
-		out, err := cmd.CombinedOutput()
-		if err != nil {
-			return fmt.Errorf("reloading bluetooth service after config update: %w, output: %s", err, out)
-		}
+	n.logger.Debug("setting bluetooth to discoverable")
+	err = adapter.SetProperty("org.bluez.Adapter1.Discoverable", dbus.MakeVariant(true))
+	if err != nil {
+		return errw.Wrap(err, "enabling bluetooth discovery")
+	}
+
+	discoveryTimeout := uint32(time.Duration(n.Config().RetryConnectionTimeoutMinutes * 2).Seconds())
+	err = adapter.SetProperty("org.bluez.Adapter1.DiscoverableTimeout", dbus.MakeVariant(discoveryTimeout))
+	if err != nil {
+		return errw.Wrap(err, "adjusting discovery timeout")
+	}
+
+	if err := conn.Export(&basicAgent{logger: n.logger, conn: conn}, BluezAgentPath, BluezAgent); err != nil {
+		return errw.Wrap(err, "exporting custom agent object")
 	}
 
+	obj := conn.Object(BluezDBusService, "/org/bluez")
+	call := obj.Call("org.bluez.AgentManager1.RegisterAgent", 0, dbus.ObjectPath(BluezAgentPath), "NoInputNoOutput")
+	if err := call.Err; err != nil {
+		return errw.Wrap(err, "registering custom agent")
+	}
+
+	call = obj.Call("org.bluez.AgentManager1.RequestDefaultAgent", 0, dbus.ObjectPath(BluezAgentPath))
```

Comment:
[q] Any reason to make your `NoInputNoOutput` agent the _default_ agent on the device? Were you having issues without this?

Reply:
The normal "register agent" functions in other libraries all do this. The idea was maybe I could hook the incoming requests from the pop-up. Otherwise, the registered agent is supposed to ONLY deal with pairing requests our app initiates (and we don't initiate pairing requests ourselves.) Feel free to play around with it.